torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -15,22 +12,255 @@ namespace datasets {
 /// A chunk could be an entire file, such as an audio data file or an image,
 /// or part of a file in the case of a large text-file split based on seek
 /// positions.
-template <typename Self, typename Batch = std::vector<Example<>>>
+template <typename Chunk = std::vector<Example<>>>
 class ChunkDataReader {
  public:
-  using SelfType = Self;
-  using BatchType = Batch;
+  using ChunkType = Chunk;
 
   /// Read an entire chunk.
-  virtual Batch read_chunk(size_t chunk_index) = 0;
+  virtual ChunkType read_chunk(size_t chunk_index) = 0;
 
   /// Returns the number of chunks available in this reader.
-  virtual size_t get_chunk_count() = 0;
+  virtual size_t chunk_count() = 0;
 
   /// This will clear any internal state associate with this reader.
   virtual void reset() = 0;
 };
 
+namespace detail {
+/// BatchDataBuffer manages a queue of UnwrappedBatchData. After a new chunk is
+/// loaded, BatchDataBuffer splits it into small batches and push them into the
+/// queue. When get_batch is called from data loader, it pops cached batches and
+/// return. If the cache is empty, it either waits to load more chunks or return
+/// null if all chunks are loaded.
+template <
+    typename UnwrappedBatch = std::vector<Example<>>,
+    typename ExampleSampler = samplers::RandomSampler>
+class BatchDataBuffer {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+  using BatchType = torch::optional<UnwrappedBatchType>;
+  using BatchRequestType = typename ExampleSampler::BatchRequestType;
+
+  BatchDataBuffer(
+      size_t num_chunks,
+      size_t batch_size,
+      ExampleSampler example_sampler,
+      bool ignore_empty_chunk,
+      size_t cache_size)
+      : remaining_chunk_count_(num_chunks),
+        batch_size_(batch_size),
+        example_sampler_(std::move(example_sampler)),
+        ignore_empty_chunk_(ignore_empty_chunk),
+        queue_capacity_(cache_size) {}
```

Comment:
let's use the same names for these. I like the refill threshold which clearly indicate what happens

Reply:
I feel refill_threshold somehow gives a sense of "this is the threshold to refill", which means to refill when reaching the threshold. This is actually the opposite of what it is doing here -- stop refill when reach the threshold. What do you think?

Reply:
I am fine with the name, just make them equal.